File: java/src/day08/HauntedWasteland.java
package day08;

import java.io.IOException;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

public class HauntedWasteland {
    private static List<String> directions = new ArrayList<>();
    private static String going = "";
    private static final Map<String, List<String>> instructions = new HashMap<>();

    public static void main(String[] args) {
        try {
            processFile("inputs/day08/input.txt");
            System.out.println("Part 1: " + part1());
            System.out.println("Part 2: " + part2());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static long part1() {
        long result = 0;
        going = "AAA";
        while (!going.equals("ZZZ")) {
            List<String> instruction = instructions.get(going);
            result = followInstructions(instruction, result);
        }
        return result;
    }

    private static long part2() {
        List<String> positions = new ArrayList<>();
        for (String key : instructions.keySet()) {
            if (key.endsWith("A")) {
                positions.add(key);
            }
        }
        Integer dirSize = directions.size();
        Map<String, Long> seen = new HashMap<>();
        return findCycle(positions, seen, dirSize);
    }

    private static long findCycle(List<String> positions, Map<String, Long> seen, Integer dirSize) {
        Map<String, Long> cycle = new HashMap<>();
        for (String position : positions) {
            List<String> instruction = instructions.get(position);
            long result = 0;
            while (seen.get(position) == null || seen.get(position) != result % dirSize) {
                if (directions.get((int) (result % dirSize)).equals("L")) {
                    position = instruction.get(0);
                } else {
                    position = instruction.get(1);
                }
                result += 1;
                if (position.endsWith("Z"))
                    seen.put(position, result % dirSize);
                instruction = instructions.get(position);
            }
            cycle.put(position, result);
        }
        Long finalRes = 1L;
        for (Long value : cycle.values()) {
            finalRes = lcm(finalRes, value);
        }
        return finalRes;
    }

    private static long gcd(long a, long b) {
        while (b > 0) {
            long temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    private static long lcm(long a, long b) {
        return a * (b / gcd(a, b));
    }

    private static long followInstructions(List<String> instruction, long result) {
        for (String direction : directions) {
            if (direction.equals("L")) {
                going = instruction.get(0);
            } else {
                going = instruction.get(1);
            }
            result += 1;
            if (going.equals("ZZZ")) {
                return result;
            }
            instruction = instructions.get(going);
        }
        return result;
    }

    private static void processFile(String file) throws IOException {
        Path currentWorkingDir = FileSystems.getDefault().getPath(System.getProperty("user.dir"));
        Path absoluteFilePath = currentWorkingDir.resolve(file);
        String content = new String(Files.readAllBytes(absoluteFilePath));
        String[] separateDirFromIns = content.strip().split("\n\n");
        directions = Arrays.stream(separateDirFromIns[0].strip().split("")).toList();

        List<String> ins = Arrays.stream(separateDirFromIns[1].strip().split("\n")).toList();

        for (String line : ins) {
            String[] insProc = line.strip().split(" = ");
            String key = insProc[0];
            List<String> value =
                    Arrays.stream(insProc[1].replace("(", " ").replace(")", " ")
                            .strip().split(", ")).toList();
            instructions.put(key, value);
        }
    }
}

File: type-script/day08/haunted-wasteland.ts
import * as fs from "fs";

const f: string = fs.readFileSync("../inputs/day08/input.txt", { encoding: "utf8" });
const fullInstructuins: string[] = f.split("\n\n");
const direction: string[] = fullInstructuins[0].trim().split("");
const network: string[][] = fullInstructuins[1].split("\n").map(x => x.trim().split("=").map(y => y.trim()));
const instructins: Map<string, string[]> = new Map(network.map(x => [x[0], x[1].replace("(", " ").replace(")", " ").trim().split(", ")]))

function part1(): number {
    let result: number = 0;
    let going: string = "AAA";
    while (going !== "ZZZ") {
        going = followPath(going, result);
        result++;
    }
    return result;
}

function part2(): number {
    const postions: string[] = [];
    for (let key of instructins.keys()) {
        if (key.endsWith("A")) {
            postions.push(key);
        }
    }
    const seen: Map<string, number> = new Map();
    return findCycle(seen, postions);
}

function followPath(going: string, result: number) {
    const ins: string[] = instructins.get(going) || [];
    const dir: string = direction[result % direction.length];
    going = getDirection(dir, going, ins);
    return going;
}

function findCycle(seen: Map<string, number>, postions: string[]) {
    const cycles: number[] = [];
    for (let position of postions) {
        let result: number = 0;
        while (seen.get(position) === undefined || seen.get(position) != result % direction.length) {
            const ins: string[] = instructins.get(position) || [];
            if (direction[result % direction.length] === "L") {
                position = ins[0];
            } else {
                position = ins[1];
            }
            result += 1;
            if (position.endsWith("Z")) {
                seen.set(position, result % direction.length);
            }
        }
        cycles.push(result);
    }
    let finallResult: number = 1;
    for (const res of cycles) {
        finallResult = lcm(res, finallResult);
    }

    return finallResult;
}

function getDirection(dir: string, going: string, ins: string[]) {
    if (dir === "L") {
        going = ins[0];
    } else {
        going = ins[1];
    }
    return going;
}

function gcd(a: number, b: number): number {
    while (b > 0) {
        const temp: number = b;
        b = a % b;
        a = temp;
    }
    return a;
}

function lcm(a: number, b: number): number {
    return a * (b / gcd(a, b));
}


console.log("Part 1:", part1());
console.log("Part 2:", part2());
